### PR TITLE
Slim Down

### PR DESCRIPTION
**Pull request by mgreystone**
I'd like to use authentic in a lambda@edge. Unfortunately, there are strict size limits, & authentic would need to be slimmed down significantly. This is my proposal:

* Bump to latest `jwks-rsa`, which drops `request` in favor of `axios` & drops `lodash`
* Drop `gimme` in favor of `axios`, since `jwks-rsa` is using it already
* Drop `funky`
* Drop `ramda`

Misc
* Add node 12 travis while dropping node 6

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,64 +1,65 @@
-const Boom     = require('boom')
-const gimme    = require('@articulate/gimme')
-const jwks     = require('jwks-rsa')
-const jwt      = require('jsonwebtoken')
+const axios                 = require('axios')
+const deepmerge             = require('deepmerge')
```
Comment (chrismcleod) [inline index.js]: this module provides the same functionality and is 7kB vs 30kB.  I tested in run kit and it gives the same results for deepmerge's example.

https://www.npmjs.com/package/@ianwalter/merge

also I only saw one use of deepmerge and it seems like a basic spread would suffice

File: index.js [hunk 2]
```diff
@@ -1,64 +1,65 @@
-const Boom     = require('boom')
-const gimme    = require('@articulate/gimme')
-const jwks     = require('jwks-rsa')
-const jwt      = require('jsonwebtoken')
+const axios                 = require('axios')
```
Comment (chrismcleod) [inline index.js:1]: https://www.npmjs.com/package/node-fetch is ~50% smaller than axios

File: index.js [hunk 3]
```diff
@@ -1,64 +1,65 @@
...
 
-const stripBearer =
-  replace(/^Bearer /i, '')
+const stripBearer = str =>
+  str.replace(/^Bearer /i, '')
 
-const throwIfNull =
-  when(isNil, () => reject('invalid token'))
+const throwIfNull = val =>
+  val == null ? Promise.reject('invalid token') : val
```
Comment (chrismcleod) [inline index.js:38]:
```suggestion
  val === null ? Promise.reject('invalid token') : val
```

File: index.js [hunk 4]
```diff
@@ -1,64 +1,65 @@
...
+
+const verifyP =
+  promisify(jwt.verify)
 
 const jwksOptsDefaults = { jwks: { cache: true, rateLimit: true } }
 
 const factory = options => {
   const clients = {}
-  const opts = mergeDeepRight(jwksOptsDefaults, options)
+  const opts = deepmerge(jwksOptsDefaults, options)
```
Comment (chrismcleod) [inline index.js]: Object.assign?  or spreads?

**Review by chrismcleod — APPROVED** [review]
Nothing broken, only suggestions to make it even smaller.

File: index.js (hunk 3 above)
Comment (mgreystone) [reply index.js:38]: Soft == for "nil"--match `null` or `undefined` but not anything else falsy

File: index.js (hunk 4 above)
Comment (mgreystone) [reply index.js]: Can't do a deep merge with either. I could make my own, of course, but i'd just assume just use a package.

File: index.js (hunk 1 above)
Comment (mgreystone) [reply index.js]: 👍 Smaller is good.

File: index.js (hunk 2 above)
Comment (mgreystone) [reply index.js:1]: `jwks-rsa` already brings in `axios`. Otherwise, i would have just went native `https`.

File: index.js (hunk 3 above)
Comment (chrismcleod) [reply index.js:38]: yea js is crazy, i suggested to favor unambiguous strict equality vs relying on the idiosyncrasies in js, all good though nbd.

File: index.js (hunk 4 above)
Comment (chrismcleod) [reply index.js]: i suppose options can be a crazy object.  if not i was thinking if smaller is better might as well do something like:

```javascript
opts = {...jwksOptsDefaults, ...options, jwks: {...jwksOptsDefaults.jwks, ...options.jwks}, someOtherOption: {...options.someOtherOption}}
```

**Review by tylerodonnell — APPROVED** [review]
Everything looks good from an infosec perspective. I'm able to confirm that this PR doesn't change any functionality with how it previously worked. Nice job.

File: index.js (hunk 4 above)
Comment (mgreystone) [reply index.js]: I guess that's true. I didn't really look at what we were merging...

File: index.js [hunk 5]
```diff
@@ -94,7 +99,12 @@ const factory = options => {
       .then(verify(token))
       .catch(deny)
 
-  return composeP(authentic, stripBearer, tapP(enforce))
+  return token =>
+    Promise.resolve(token)
+      .then(enforce)
+      .then(() => token)
```
Comment (kdstew) [inline index.js]: Could use the `tapP` added in `lib/helpers` here.

**Review by kdstew — APPROVED** [review]
One small comment otherwise looks good

**Review by jrpz — APPROVED** [review]

**Review by spencerfdavis — APPROVED** [review]

**Review by dpbackes — APPROVED** [review]
I'll give this the ol' :stamp: on behalf of 360-web. As we start to roll this out, are there particular things we should have QA look for?

**Comment by mgreystone** [comment]
> As we start to roll this out, are there particular things we should have QA look for?

Not really. Just regressions, i think.

**Review by flintinatux — APPROVED** [review]
I'm always down for adding Lambda support.  The only suggestion/thought I'd have is that you can replace much of Ramda with [tinyfunk](https://github.com/flintinatux/tinyfunk).  You'd get to keep the functional, but seriously cut down on file-size.  Totally just a thought, though, since I see you've solved things in more imperative ways in some places.

**Comment by flintinatux** [comment]
@mgreystone, one more question:  If the concern is for Lambda package size limits, have you considered using Lambda Layers?

**Comment by spencerfdavis** [comment]
image

**Comment by mgreystone** [comment]
> @mgreystone, one more question: If the concern is for Lambda package size limits, have you considered using Lambda Layers?

I have not. I'm not sure i follow how i'd use layers to solve this problem? Especially in a lambda@edge? This _does_ shave just shy of 2 MB the size of my lambda package, but i'm honestly not confident it's going to be enough...

**Comment by flintinatux** [comment]
The idea with layers is that you can put all of your deps in a layer, and your actual lambda code in the function itself.  Each layer can be up to the max zip-file size supported by Lambda, and you can have up to 5 layers per function.  So it increases the size limit of your uploads.

I've built Lambda's before where all the node modules were in a layer, and then the lambda function was just a single file.  It's not a one-size-fits-all solution, and I'm not sure if our Jenkins flows supports it, but layers can help with upload size.

**Pull request merged**